src/main/java/codeu/model/store/basic/MessageStore.java

```diff
@@ -56,37 +59,115 @@ public static MessageStore getTestInstance(PersistentStorageAgent persistentStor
    */
   private PersistentStorageAgent persistentStorageAgent;
 
-  /** The in-memory list of Messages. */
-  private List<Message> messages;
+  /** The in-memory map of Messages. */
+  private Map<UUID, Message> messages;
+
+  // These map from some conversation/user ID to a list containing the IDs of their messages
+  private Map<UUID, List<UUID>> conversationIdToMessageIdList;
+  private Map<UUID, List<UUID>> authorIdToMessageIdList;
 
   /** This class is a singleton, so its constructor is private. Call getInstance() instead. */
   private MessageStore(PersistentStorageAgent persistentStorageAgent) {
     this.persistentStorageAgent = persistentStorageAgent;
-    messages = new ArrayList<>();
+    messages = new HashMap<>();
+
+    conversationIdToMessageIdList = new HashMap<>();
+    authorIdToMessageIdList = new HashMap<>();
   }
 
-  /** Add a new message to the current set of messages known to the application. */
+  /**
+   * Add a new message to the current set of messages known to the application.
+   * This writes the message to the persistent storage.
+   */
   public void addMessage(Message message) {
-    messages.add(message);
+    addMessageWithoutPersistentStorage(message);
+
     persistentStorageAgent.writeThrough(message);
   }
 
+  /**
+   * Add a new message to the current set of messages known to the application.
+   * This does NOT write the message to the persistent storage.
+   */
+  public void addMessageWithoutPersistentStorage(Message message) {
+    UUID id = message.getId();
+    UUID conversationId = message.getConversationId();
+    UUID authorId = message.getAuthorId();
+
+    messages.put(id, message);
+
+    // Add to its conversation's list
+    if (conversationIdToMessageIdList.containsKey(conversationId)) {
+      conversationIdToMessageIdList.get(conversationId).add(id);
+    } else {
+      conversationIdToMessageIdList.put(conversationId, new ArrayList<>());
+      conversationIdToMessageIdList.get(conversationId).add(id);
+    }
+
+    // Add to its author's list
+    if (authorIdToMessageIdList.containsKey(authorId)) {
+      conversationIdToMessageIdList.get(authorId).add(id);
+    } else {
+      conversationIdToMessageIdList.put(authorId, new ArrayList<>());
+      conversationIdToMessageIdList.get(authorId).add(id);
+    }
+  }
+
   /** Access the current set of Messages within the given Conversation. */
   public List<Message> getMessagesInConversation(UUID conversationId) {
 
-    List<Message> messagesInConversation = new ArrayList<>();
+    List<Message> messagesInConversation;
 
-    for (Message message : messages) {
-      if (message.getConversationId().equals(conversationId)) {
-        messagesInConversation.add(message);
-      }
+    if (conversationIdToMessageIdList.containsKey(conversationId)) {
+      messagesInConversation = // To get the messages in a conversation,
```

Comment:
nit: I think a lot of the comments here are redundant, and the extra whitespace makes things harder to read/understand not easier.

Imo if you had one top-level comment along the lines of "Get all messages from conversation id" it would be fine.

Reply:
Done.